backend/src/schema/resolvers/images/images.spec.js

```diff
@@ -0,0 +1,344 @@
+import { deleteImage, mergeImage } from './images'
+import { getNeode, getDriver } from '../../../db/neo4j'
+import Factory, { cleanDatabase } from '../../../db/factories'
+import { UserInputError } from 'apollo-server'
+
+const driver = getDriver()
+const neode = getNeode()
+const uuid = '[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}'
+let uploadCallback
+let deleteCallback
+
+beforeEach(async () => {
+  await cleanDatabase()
+  uploadCallback = jest.fn(({ destination }) => destination)
+  deleteCallback = jest.fn()
+})
+
+describe('deleteImage', () => {
+  describe('given a resource with an image', () => {
+    let user
+    beforeEach(async () => {
+      user = await Factory.build(
+        'user',
+        {},
+        {
+          avatar: Factory.build('image', {
+            url: '/some/avatar/url/',
+            alt: 'This is the avatar image of a user',
+          }),
+        },
+      )
+      user = await user.toJson()
+    })
+
+    it('soft deletes `Image` node', async () => {
+      await expect(neode.all('Image')).resolves.toHaveLength(1)
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      await expect(neode.all('Image')).resolves.toHaveLength(0)
+    })
+
+    it('calls deleteCallback', async () => {
+      user = await Factory.build('user')
+      user = await user.toJson()
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      expect(deleteCallback).toHaveBeenCalled()
+    })
+
+    describe('given a transaction parameter', () => {
+      it('executes cypher statements within the transaction', async () => {
+        const session = driver.session()
+        let someString
+        try {
+          someString = await session.writeTransaction(async transaction => {
+            await deleteImage(user, 'AVATAR_IMAGE', {
+              deleteCallback,
+              transaction,
+            })
+            const txResult = await transaction.run('RETURN "Hello" as result')
+            const [result] = txResult.records.map(record => record.get('result'))
+            return result
+          })
+        } finally {
+          session.close()
+        }
+        await expect(neode.all('Image')).resolves.toHaveLength(0)
+        await expect(someString).toEqual('Hello')
+      })
+
+      it('rolls back the transaction in case of errors', async done => {
+        await expect(neode.all('Image')).resolves.toHaveLength(1)
+        const session = driver.session()
+        try {
+          await session.writeTransaction(async transaction => {
+            await deleteImage(user, 'AVATAR_IMAGE', {
+              deleteCallback,
+              transaction,
+            })
+            throw new Error('Ouch!')
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-26T08:39:48Z" title="Wednesday, February 26th 2020, 9:39:48 am +01:00">Feb 26, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,347 @@
+import { deleteImage, mergeImage } from './images'
+import { getNeode, getDriver } from '../../../db/neo4j'
+import Factory, { cleanDatabase } from '../../../db/factories'
+import { UserInputError } from 'apollo-server'
+
+const driver = getDriver()
+const neode = getNeode()
+const uuid = '[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}'
+let uploadCallback
+let deleteCallback
+
+beforeEach(async () => {
+  await cleanDatabase()
+  uploadCallback = jest.fn(({ destination }) => destination)
+  deleteCallback = jest.fn()
+})
+
+describe('deleteImage', () => {
+  describe('give a resource with an image', () => {
+    let user
+    beforeEach(async () => {
+      user = await Factory.build(
+        'user',
+        {},
+        {
+          avatar: Factory.build('image', {
+            url: '/some/avatar/url/',
+            alt: 'This is the avatar image of a user',
+          }),
+        },
+      )
+      user = await user.toJson()
+    })
+
+    it('soft deletes `Image` node', async () => {
+      await expect(neode.all('Image')).resolves.toHaveLength(1)
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      await expect(neode.all('Image')).resolves.toHaveLength(0)
+    })
+
+    it('calls deleteCallback', async () => {
+      user = await Factory.build('user')
+      user = await user.toJson()
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      expect(deleteCallback).toHaveBeenCalled()
+    })
+
+    describe('given a transaction parameter', () => {
+      it('executes cypher statements within the transaction', async () => {
+        const session = driver.session()
+        let someString
+        try {
+          someString = await session.writeTransaction(async transaction => {
+            await deleteImage(user, 'AVATAR_IMAGE', {
+              deleteCallback,
+              transaction,
+            })
+            const txResult = await transaction.run('RETURN "Hello" as result')
+            const [result] = txResult.records.map(record => record.get('result'))
+            return result
+          })
+        } finally {
+          session.close()
+        }
+        await expect(neode.all('Image')).resolves.toHaveLength(0)
+        await expect(someString).toEqual('Hello')
+      })
+
+      it('rolls back the transaction in case of errors', async done => {
+        await expect(neode.all('Image')).resolves.toHaveLength(1)
+        const session = driver.session()
+        try {
+          await session.writeTransaction(async transaction => {
+            await deleteImage(user, 'AVATAR_IMAGE', {
+              deleteCallback,
+              transaction,
+            })
+            throw new Error('Ouch!')
```


I lost my first comment on this, cause of issues with GitHub, but then we talked in person about this.

If I understood correctly, what is happening in these tests is that sometimes the `deleteImage` function is called in a "bulk" delete where a user might want to delete all their posts, as well as their account. Inside that transaction, we must delete images as well. Assuming the `deleteImage` does not throw and error, but some other transaction does, then we want to check that the `images` are not deleted, but that the whole transaction fails. 
As you put it yesterday, it's all or nothing. To achieve this in a test, you are just throwing an error, and checking the state of the database.